### ISSUE TITLE
[Scala 2.13] Subtype of Newtype cause cyclic subtyping error 

### ISSUE THREAD

**Issue by guizmaii** [zio/zio-prelude#1183]
Same issue as in https://github.com/zio/zio-prelude/issues/817

Reproducer:
```scala
object type {
  type UUIDv1 = UUIDv1.Type
  object UUIDv1 extends Subtype[UUID]
  
  type UserId = UserId.Type
  object UserId extends Subtype[UUIDv1]
}
```

Error:
```scala
[error] example.scala:48:8: cyclic aliasing or subtyping involving type Type
[error]   type UserId = UserId.Type
```

Cc @kitlangton

**Comment by guizmaii** [comment]
See also: https://github.com/zio/zio-prelude/pull/949#issuecomment-1201735998

**Comment by cheikh2shift** [comment]
/attempt #1183

<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clnj2g30q0002ml0fa0440m4j/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by guptasamarth61** [comment]
/attempt #1183

<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clnj2g30q0002ml0fa0440m4j/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by Tan28bha** [comment]
/attempt #1183

<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clnj2g30q0002ml0fa0440m4j/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by Tan28bha** [comment]
[open source.txt](https://github.com/zio/zio-prelude/files/13254211/open.source.txt)
 /claim #1183

**Comment by abebeos** [comment]
@guizmaii , could you please provide a self-contained reproducer that works independently of zio, right from the cli via a usual `scalac` command?

**Comment by abebeos** [comment]
@ team: this joke of a bounty-system does not allow me to cancel my attempt. Please do so from your side)

> The issue is to fix the implementation in ZIO Prelude so that users do not have to use a private compiler flag as a workaround.
from https://github.com/zio/zio-prelude/pull/1228#issuecomment-1824661085

To whomever attempts this:

"fixing" this within zio-prelude is essentially "plumbing around". The fix need to be within scalac itself.

**Comment by abebeos** [comment]
@guizmaii 

You write "Same issue as in https://github.com/zio/zio-prelude/issues/817", but your reproducer is different.

Issue #817 reproducer

```scala
// ---------------------------------------------------------------------------
// Subtype of Newtype cause cyclic subtyping error
// https://github.com/zio/zio-prelude/issues/817

object GenericItemId extends Newtype[UUID]
type GenericItemId = GenericItemId.Type

object SpecificItemId extends Subtype[GenericItemId]
type SpecificItemId = SpecificItemId.Type
```

Your reproducer

```scala
// ---------------------------------------------------------------------------
// [Scala 2.13] Subtype of Newtype cause cyclic subtyping error
// https://github.com/zio/zio-prelude/issues/1183

type UUIDv1 = UUIDv1.Type
object UUIDv1 extends Subtype[UUID]
  
type UserId = UserId.Type
object UserId extends Subtype[UUIDv1]
```

Your reproducer does not use `Newtype`.

If your rewrite your reproducer to use `Newtype` instead of first `Subtype`, things work fine

```scala
type UUIDv1 = UUIDv1.Type
object UUIDv1 extends Newtype[UUID]
//                    ^^^
type UserId = UserId.Type
object UserId extends Subtype[UUIDv1]
```

A further rewrite to make your reproducer similar to the one from https://github.com/zio/zio-prelude/issues/817

```scala
object UUIDv1 extends Newtype[UUID]
type UUIDv1 = UUIDv1.Type

object UserId extends Subtype[UUIDv1]
type UserId = UserId.Type
```

### Non Issue

@adamgfraser , to me it looks like this here is a non-issue (not a bug in zio-prelude, but a bug in application/reproducer code)

**Comment by adamgfraser** [comment]
@abebeos This is wrong. Users should be able to define types that are subtypes or new types of other user defined types, whether those types are subtypes or new types.

**Comment by abebeos** [comment]
> This is wrong. Users should be able to define types that are subtypes or new types of other user defined types, whether those types are subtypes or new types.

the get your test-cases/reproducers straight, and document them clearly.

"Subtype of Subtype" is NOT the "Same issue" (as stated by the OP) with "Subtype of Newtype".

**Comment by abebeos** [comment]
> This is wrong.

Take a look at the title: "Subtype of Newtype cause cyclic subtyping error"

The reproducer is not about "Subtype of Newtype", but about "Subtype of Subtype".

So, in any case, either the issue-title or the reproducer is bugged (and for sure the statement "Same issue" is wrong).

**Comment by abebeos** [comment]
@adamgfraser , @guizmaii , if you're incapable to accept `Subtype != Newtype`, then simply verify it with the older issue

Pick https://github.com/zio/zio-prelude/issues/817, Change Newtype to Subtype, compile with 1.0.0-RC7 => Fails

```
import zio.prelude._
import java.util.UUID

// works with RC7, fails with RC8
object zio_prelude_i817_Newtype {
 object GenericItemId extends Newtype[UUID]
 type GenericItemId = GenericItemId.Type

 object SpecificItemId extends Subtype[GenericItemId]
 type SpecificItemId = SpecificItemId.Type
}

// fails with RC7, fails with RC8
object zio_prelude_i817_Subtype {
 object GenericItemId extends Subtype[UUID]
 type GenericItemId = GenericItemId.Type

 object SpecificItemId extends Subtype[GenericItemId]
 type SpecificItemId = SpecificItemId.Type
}
```

Why? Because `Subtype != Newtype` (they have different behavior, in addition to being different words).

It is quite fascinating how your domain-knowledge makes you incapable to see this simple fact.

**Comment by Krishnakalani111** [comment]
/attempt #1183

<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clnj2g30q0002ml0fa0440m4j/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by SethTisue** [comment]
(I'm not certain it's relevant as I haven't followed the history closely, but regardless, FYI:)

a fix for https://github.com/scala/bug/issues/12622 has been merged for Scala 2.13.13 and is available for testing in [Scala 2 nightly builds](https://stackoverflow.com/q/40622878/86485)

**Comment by abebeos** [comment]
> I'm not certain it's relevant

it should enable 'subtype of subtype' (though this issue is still about 'subtype of newtype', according to the title).

The Precision of Words...

**Comment by adamgfraser** [comment]
@SethTisue Awesome! I can confirm that this resolves the issue. Do you have any idea when the next version will be released? Thanks for the update! 😃

**Comment by adamgfraser** [comment]
This will be fixed when the next Scala 2.13 version is released.

**Comment by SethTisue** [comment]
probably January — you can monitor https://contributors.scala-lang.org/t/scala-2-13-13-release-planning/6315 if you like
